### PR TITLE
pessimistic-txn: support deadlock detector in mocktikv

### PR DESCRIPTION
**Pull request by coocood**
### What problem does this PR solve? 
Support deadlock detector in mocktikv for pessimistic transaction.

### What is changed and how it works?
add 'util/deadlock' package and use it in mocktikv.

### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: store/mockstore/mocktikv/mvcc_leveldb.go [hunk 1]
```diff
@@ -860,13 +873,17 @@ func getTxnCommitInfo(iter *Iterator, expectKey []byte, startTS uint64) (mvccVal
...
+		mvcc.mu.Unlock()
+		mvcc.deadlockDetector.CleanUp(startTS)
+	}()
 
 	batch := &leveldb.Batch{}
 	err := rollbackKey(mvcc.db, batch, key, startTS)
 	if err != nil {
 		return errors.Trace(err)
 	}
+	mvcc.deadlockDetector.CleanUp(startTS)
```
Comment (tiancaiamao) [inline store/mockstore/mocktikv/mvcc_leveldb.go]: Redundant?  Line 878 already do it

File: util/deadlock/deadlock.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+		if len(l.txns) == 0 {
+			delete(d.waitForMap, txn)
+		}
+	}
+	d.lock.Unlock()
+
+}
+
+// Expire removes entries with TS smaller than minTS.
+func (d *Detector) Expire(minTS uint64) {
```
Comment (tiancaiamao) [inline util/deadlock/deadlock.go]: It's not actually used?

File: session/pessimistic_test.go [hunk 3]
```diff
@@ -165,3 +168,29 @@ func (s *testPessimisticSuite) TestTxnMode(c *C) {
...
+		tk1.MustExec("begin pessimistic")
+		tk1.MustExec("update deadlock set v = v + 1 where k = 2")
+		<-syncCh
+		tk1.MustExec("update deadlock set v = v + 1 where k = 1")
+		<-syncCh
+	}()
+	tk.MustExec("begin pessimistic")
+	tk.MustExec("update deadlock set v = v + 1 where k = 1")
+	syncCh <- struct{}{}
+	time.Sleep(time.Millisecond * 10)
```
[the remark this answers is not in the document]
Comment (coocood) [reply session/pessimistic_test.go]: Then there would be Resolve Lock timeout error.

File: session/pessimistic_test.go [hunk 4]
```diff
@@ -165,3 +168,29 @@ func (s *testPessimisticSuite) TestTxnMode(c *C) {
...
+	_, err := tk.Exec("update deadlock set v = v + 1 where k = 2")
+	e, ok := errors.Cause(err).(*terror.Error)
+	c.Assert(ok, IsTrue)
+	c.Assert(int(e.Code()), Equals, mysql.ErrLockDeadlock)
+	syncCh <- struct{}{}
```
[the remark this answers is not in the document]
Comment (coocood) [reply session/pessimistic_test.go]: It ensures that the tk1 successfully execute the last statement.

File: util/deadlock/deadlock.go (hunk 2 above)
Comment (coocood) [reply util/deadlock/deadlock.go]: Will be used in the future.

File: store/mockstore/mocktikv/mvcc_leveldb.go (hunk 1 above)
Comment (coocood) [reply store/mockstore/mocktikv/mvcc_leveldb.go]: yes

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**